File: evaluation/results/scenario_2/r/claude-sonnet-4-20250514/run_03/extracted_code.R
# Load required libraries
library(rstan)
library(dplyr)
library(ggplot2)
library(tidyr)
library(lubridate)

# Stan model code
stan_model_code <- "
data {
  int<lower=1> T;                    // Number of days
  int<lower=0> cases[T];             // Reported cases
  int<lower=1,upper=7> day_of_week[T]; // Day of week (1=Monday, 7=Sunday)
  int<lower=1> S;                    // Maximum generation interval
  vector[S] gen_interval;            // Generation interval PMF
  int<lower=1> D;                    // Maximum delay
  vector[D] delay_pmf;               // Delay from infection to reporting PMF
  int<lower=1> seed_days;            // Number of initial seeding days
}

transformed data {
  int effective_T = T - seed_days;   // Days after seeding period
}

parameters {
  // Initial infections (seeding period)
  vector<lower=0>[seed_days] log_infections_seed;
  
  // Rt parameters
  real<lower=0> R_initial;           // Initial Rt
  vector[effective_T-1] rt_innovations; // Random walk innovations
  real<lower=0> rt_sigma;            // Random walk standard deviation
  
  // Day-of-week effects (multiplicative, sum to 7)
  vector[6] dow_raw;                 // 6 free parameters for 7 effects
  
  // Time-varying ascertainment
  real logit_asc_initial;            // Initial ascertainment (logit scale)
  vector[effective_T-1] asc_innovations; // Random walk innovations
  real<lower=0> asc_sigma;           // Ascertainment random walk SD
  
  // Overdispersion
  real<lower=0> phi;                 // Negative binomial overdispersion
}

transformed parameters {
  vector[T] log_infections;
  vector[T] infections;
  vector[T] expected_cases;
  vector[7] dow_effects;
  vector[effective_T] rt;
  vector[effective_T] ascertainment;
  
  // Seeding period infections
  log_infections[1:seed_days] = log_infections_seed;
  
  // Day-of-week effects (constraint: sum = 7)
  dow_effects[1:6] = dow_raw;
  dow_effects[7] = 7 - sum(dow_raw);
  
  // Rt (random walk on log scale)
  rt[1] = R_initial;
  for(t in 2:effective_T) {
    rt[t] = rt[t-1] * exp(rt_innovations[t-1]);
  }
  
  // Ascertainment (random walk on logit scale)
  ascertainment[1] = inv_logit(logit_asc_initial);
  for(t in 2:effective_T) {
    ascertainment[t] = inv_logit(logit(ascertainment[t-1]) + asc_innovations[t-1]);
  }
  
  // Renewal equation for infections after seeding
  for(t in (seed_days + 1):T) {
    real convolution = 0;
    int rt_idx = t - seed_days;
    
    for(s in 1:min(S, t-1)) {
      convolution += infections[t-s] * gen_interval[s];
    }
    
    log_infections[t] = log(rt[rt_idx] * convolution + 1e-10);
  }
  
  // Convert to natural scale
  infections = exp(log_infections);
  
  // Expected reported cases (with delay and day-of-week effects)
  for(t in 1:T) {
    real delayed_infections = 0;
    
    // Apply delay from infection to reporting
    for(d in 1:min(D, t)) {
      int inf_day = t - d + 1;
      if(inf_day >= 1) {
        delayed_infections += infections[inf_day] * delay_pmf[d];
      }
    }
    
    // Apply ascertainment (use last available value for seeding period)
    real current_asc;
    if(t <= seed_days) {
      current_asc = ascertainment[1];
    } else {
      current_asc = ascertainment[t - seed_days];
    }
    
    expected_cases[t] = delayed_infections * current_asc * dow_effects[day_of_week[t]];
  }
}

model {
  // Priors
  log_infections_seed ~ normal(5, 2);   // Initial infections
  R_initial ~ normal(1, 0.5);
  rt_innovations ~ normal(0, rt_sigma);
  rt_sigma ~ exponential(20);           // Promotes smoothness
  
  dow_raw ~ normal(1, 0.2);             // Day-of-week effects around 1
  
  logit_asc_initial ~ normal(-2, 1);    // Initial ascertainment around 0.1
  asc_innovations ~ normal(0, asc_sigma);
  asc_sigma ~ exponential(50);          // Promotes smoothness
  
  phi ~ exponential(0.1);               // Overdispersion
  
  // Likelihood
  for(t in 1:T) {
    if(expected_cases[t] > 0) {
      cases[t] ~ neg_binomial_2(expected_cases[t], phi);
    }
  }
}

generated quantities {
  vector[T] log_lik;
  vector[T] cases_pred;
  
  for(t in 1:T) {
    if(expected_cases[t] > 0) {
      log_lik[t] = neg_binomial_2_lpmf(cases[t] | expected_cases[t], phi);
      cases_pred[t] = neg_binomial_2_rng(expected_cases[t], phi);
    } else {
      log_lik[t] = 0;
      cases_pred[t] = 0;
    }
  }
}
"

# Function to create generation interval (gamma distribution)
create_generation_interval <- function(mean_gi = 5.2, sd_gi = 2.3, max_gi = 20) {
  shape <- (mean_gi / sd_gi)^2
  rate <- mean_gi / sd_gi^2
  
  gi <- dgamma(1:max_gi, shape = shape, rate = rate)
  gi / sum(gi)  # Normalize to sum to 1
}

# Function to create delay distribution (gamma distribution)
create_delay_distribution <- function(mean_delay = 8, sd_delay = 4, max_delay = 30) {
  shape <- (mean_delay / sd_delay)^2
  rate <- mean_delay / sd_delay^2
  
  delays <- dgamma(1:max_delay, shape = shape, rate = rate)
  delays / sum(delays)  # Normalize to sum to 1
}

# Function to load and prepare data
load_and_prepare_data <- function(filename) {
  # For demonstration, I'll create sample data
  # In practice, you would use: data <- read.csv(filename)
  
  dates <- seq(as.Date("2023-01-01"), as.Date("2023-06-30"), by = "day")
  
  # Simulate realistic COVID case data with patterns
  set.seed(123)
  T <- length(dates)
  
  # Simulate underlying Rt that changes over time
  true_rt <- c(
    rep(1.5, 30),           # Initial high transmission
    seq(1.5, 0.8, length.out = 60),  # Declining due to interventions
    rep(0.8, 30),           # Low transmission
    seq(0.8, 1.2, length.out = 30),  # Increase (variant or relaxed measures)
    rep(1.2, T - 150)       # Moderate transmission
  )
  
  # Simulate infections using renewal equation
  gen_interval <- create_generation_interval()
  infections <- numeric(T)
  infections[1:14] <- exp(rnorm(14, log(100), 0.3))  # Seeding
  
  for(t in 15:T) {
    convolution <- sum(infections[max(1, t-20):t-1] * rev(gen_interval[1:min(20, t-1)]))
    infections[t] <- max(1, true_rt[t] * convolution * exp(rnorm(1, 0, 0.1)))
  }
  
  # Apply delay and day-of-week effects to get reported cases
  delay_pmf <- create_delay_distribution()
  dow_effects <- c(1.0, 1.0, 1.0, 1.0, 1.0, 0.7, 0.5)  # Weekend underreporting
  ascertainment <- 0.3  # 30% of infections reported
  
  cases <- numeric(T)
  for(t in 1:T) {
    delayed_inf <- sum(infections[max(1, t-30):t] * rev(delay_pmf[1:min(30, t)]))
    dow <- wday(dates[t], week_start = 1)
    expected <- delayed_inf * ascertainment * dow_effects[dow]
    cases[t] <- rnbinom(1, mu = expected, size = 10)
  }
  
  data.frame(
    date = dates,
    cases = cases,
    day_of_week = wday(dates, week_start = 1)
  )
}

# Main estimation function
estimate_rt <- function(data, seed_days = 14) {
  # Prepare data for Stan
  gen_interval <- create_generation_interval()
  delay_pmf <- create_delay_distribution()
  
  stan_data <- list(
    T = nrow(data),
    cases = data$cases,
    day_of_week = data$day_of_week,
    S = length(gen_interval),
    gen_interval = gen_interval,
    D = length(delay_pmf),
    delay_pmf = delay_pmf,
    seed_days = seed_days
  )
  
  # Compile and fit the model
  model <- stan_model(model_code = stan_model_code)
  
  fit <- sampling(
    model,
    data = stan_data,
    chains = 4,
    iter = 2000,
    warmup = 1000,
    cores = 4,
    control = list(adapt_delta = 0.95, max_treedepth = 12)
  )
  
  return(fit)
}

# Function to extract and summarize results
extract_results <- function(fit, data, seed_days = 14) {
  # Extract posterior samples
  rt_samples <- extract(fit, "rt")$rt
  dow_samples <- extract(fit, "dow_effects")$dow_effects
  asc_samples <- extract(fit, "ascertainment")$ascertainment
  
  # Summarize Rt
  rt_summary <- data.frame(
    date = data$date[(seed_days + 1):nrow(data)],
    rt_median = apply(rt_samples, 2, median),
    rt_lower = apply(rt_samples, 2, quantile, 0.025),
    rt_upper = apply(rt_samples, 2, quantile, 0.975)
  )
  
  # Summarize day-of-week effects
  dow_summary <- data.frame(
    day_of_week = 1:7,
    day_name = c("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"),
    effect_median = apply(dow_samples, 2, median),
    effect_lower = apply(dow_samples, 2, quantile, 0.025),
    effect_upper = apply(dow_samples, 2, quantile, 0.975)
  )
  
  # Summarize ascertainment
  asc_summary <- data.frame(
    date = data$date[(seed_days + 1):nrow(data)],
    asc_median = apply(asc_samples, 2, median),
    asc_lower = apply(asc_samples, 2, quantile, 0.025),
    asc_upper = apply(asc_samples, 2, quantile, 0.975)
  )
  
  return(list(
    rt = rt_summary,
    day_of_week = dow_summary,
    ascertainment = asc_summary,
    fit = fit
  ))
}

# Function to create plots
create_plots <- function(results, data) {
  # Rt plot
  p1 <- ggplot(results$rt, aes(x = date)) +
    geom_ribbon(aes(ymin = rt_lower, ymax = rt_upper), alpha = 0.3, fill = "blue") +
    geom_line(aes(y = rt_median), color = "blue", size = 1) +
    geom_hline(yintercept = 1, linetype = "dashed", color = "red") +
    labs(title = "Time-varying Reproduction Number (Rt)",
         x = "Date", y = "Rt") +
    theme_minimal()
  
  # Day-of-week effects
  p2 <- ggplot(results$day_of_week, aes(x = day_name)) +
    geom_col(aes(y = effect_median), fill = "steelblue", alpha = 0.7) +
    geom_errorbar(aes(ymin = effect_lower, ymax = effect_upper), 
                  width = 0.2) +
    labs(title = "Day-of-Week Effects",
         x = "Day of Week", y = "Multiplicative Effect") +
    theme_minimal()
  
  # Ascertainment over time
  p3 <- ggplot(results$ascertainment, aes(x = date)) +
    geom_ribbon(aes(ymin = asc_lower, ymax = asc_upper), alpha = 0.3, fill = "green") +
    geom_line(aes(y = asc_median), color = "green", size = 1) +
    labs(title = "Time-varying Ascertainment Rate",
         x = "Date", y = "Ascertainment Rate") +
    theme_minimal()
  
  # Cases over time with model fit
  cases_pred <- extract(results$fit, "cases_pred")$cases_pred
  cases_fit <- data.frame(
    date = data$date,
    observed = data$cases,
    predicted_median = apply(cases_pred, 2, median),
    predicted_lower = apply(cases_pred, 2, quantile, 0.025),
    predicted_upper = apply(cases_pred, 2, quantile, 0.975)
  )
  
  p4 <- ggplot(cases_fit, aes(x = date)) +
    geom_ribbon(aes(ymin = predicted_lower, ymax = predicted_upper), 
                alpha = 0.3, fill = "orange") +
    geom_line(aes(y = predicted_median), color = "orange", size = 1) +
    geom_point(aes(y = observed), size = 0.5, alpha = 0.6) +
    labs(title = "Observed vs Predicted Cases",
         x = "Date", y = "Daily Cases") +
    theme_minimal()
  
  return(list(rt = p1, dow = p2, ascertainment = p3, fit = p4))
}

# Main execution
main <- function() {
  cat("Loading and preparing data...\n")
  data <- load_and_prepare_data("cases.csv")
  
  cat("Fitting model (this may take several minutes)...\n")
  fit <- estimate_rt(data)
  
  cat("Extracting results...\n")
  results <- extract_results(fit, data)
  
  cat("Creating plots...\n")
  plots <- create_plots(results, data)
  
  # Display results
  cat("\n=== RT ESTIMATES (last 10 days) ===\n")
  print(tail(results$rt, 10))
  
  cat("\n=== DAY-OF-WEEK EFFECTS ===\n")
  print(results$day_of_week)
  
  cat("\n=== ASCERTAINMENT (last 10 days) ===\n")
  print(tail(results$ascertainment, 10))
  
  # Show plots
  print(plots$rt)
  print(plots$dow)
  print(plots$ascertainment)
  print(plots$fit)
  
  # Model diagnostics
  cat("\n=== MODEL DIAGNOSTICS ===\n")
  print(summary(fit, pars = c("rt_sigma", "asc_sigma", "phi"))$summary)
  
  return(list(
    data = data,
    results = results,
    plots = plots,
    fit = fit
  ))
}

# Run the analysis
if(interactive()) {
  analysis <- main()
}

File: evaluation/results/scenario_2/pymc/claude-sonnet-4-20250514/run_01/extracted_code.py
import pandas as pd
import numpy as np
import pymc as pm
import pytensor.tensor as pt
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import gamma
import arviz as az
from datetime import datetime, timedelta
import warnings
warnings.filterwarnings('ignore')

# Set random seed for reproducibility
np.random.seed(42)

def generate_sample_data():
    """Generate sample COVID-19 case data for demonstration"""
    # Set parameters
    n_days = 150
    start_date = pd.to_datetime('2020-03-01')
    
    # True Rt trajectory (decreasing then increasing)
    t = np.linspace(0, n_days-1, n_days)
    true_rt = 2.5 * np.exp(-t/30) + 0.8 + 0.3 * np.sin(t/20)
    
    # Generation interval (Gamma distribution, mean=5.1, sd=2.3)
    max_gen_int = 20
    gen_int_shape = 4.9
    gen_int_scale = 1.04
    gen_int = gamma.pdf(np.arange(1, max_gen_int+1), a=gen_int_shape, scale=gen_int_scale)
    gen_int = gen_int / gen_int.sum()
    
    # Reporting delay (mean=7 days)
    max_delay = 25
    delay_shape = 2.0
    delay_scale = 3.5
    reporting_delay = gamma.pdf(np.arange(0, max_delay), a=delay_shape, scale=delay_scale)
    reporting_delay = reporting_delay / reporting_delay.sum()
    
    # Simulate true infections
    infections = np.zeros(n_days)
    infections[0] = 100  # Initial infections
    
    for t in range(1, n_days):
        # Renewal equation
        infectiousness = 0
        for s in range(min(t, len(gen_int))):
            infectiousness += infections[t-1-s] * gen_int[s]
        infections[t] = true_rt[t] * infectiousness
    
    # Day-of-week effects (lower reporting on weekends)
    dow_effects = np.array([1.0, 1.0, 1.0, 1.0, 1.0, 0.7, 0.5])  # Mon-Sun
    
    # Time-varying ascertainment (starts low, increases)
    ascertainment = 0.1 + 0.4 / (1 + np.exp(-(t - 50)/15))
    
    # Apply reporting delay and observation process
    expected_reports = np.zeros(n_days + max_delay)
    for t in range(n_days):
        for d in range(len(reporting_delay)):
            if t + d < len(expected_reports):
                expected_reports[t + d] += infections[t] * ascertainment[t] * reporting_delay[d]
    
    # Truncate to original period
    expected_reports = expected_reports[:n_days]
    
    # Apply day-of-week effects
    dates = pd.date_range(start_date, periods=n_days, freq='D')
    dow = dates.dayofweek  # Monday=0, Sunday=6
    dow_multiplier = dow_effects[dow]
    expected_reports *= dow_multiplier
    
    # Add overdispersion (negative binomial)
    phi = 10  # Overdispersion parameter
    cases = np.random.negative_binomial(
        n=phi, 
        p=phi / (phi + expected_reports)
    )
    
    # Create DataFrame
    df = pd.DataFrame({
        'date': dates,
        'cases': cases,
        'day_of_week': dow + 1  # Convert to 1-7 (Monday=1)
    })
    
    # Save to CSV
    df.to_csv('cases.csv', index=False)
    
    # Return additional info for validation
    return df, {
        'true_rt': true_rt,
        'true_infections': infections,
        'true_ascertainment': ascertainment,
        'true_dow_effects': dow_effects,
        'gen_int': gen_int,
        'reporting_delay': reporting_delay
    }

def create_generation_interval():
    """Create generation interval PMF"""
    # COVID-19 generation interval: Gamma(shape=4.9, scale=1.04)
    max_gen_int = 20
    gen_int_shape = 4.9
    gen_int_scale = 1.04
    
    gen_int = gamma.pdf(np.arange(1, max_gen_int+1), a=gen_int_shape, scale=gen_int_scale)
    gen_int = gen_int / gen_int.sum()
    
    return gen_int

def create_reporting_delay():
    """Create reporting delay PMF"""
    # Reporting delay: Gamma(shape=2.0, scale=3.5)
    max_delay = 25
    delay_shape = 2.0
    delay_scale = 3.5
    
    reporting_delay = gamma.pdf(np.arange(0, max_delay), a=delay_shape, scale=delay_scale)
    reporting_delay = reporting_delay / reporting_delay.sum()
    
    return reporting_delay

def load_data():
    """Load and preprocess the data"""
    try:
        df = pd.read_csv('cases.csv')
    except FileNotFoundError:
        print("cases.csv not found. Generating sample data...")
        df, true_params = generate_sample_data()
        print("Sample data generated and saved as cases.csv")
    
    df['date'] = pd.to_datetime(df['date'])
    df = df.sort_values('date').reset_index(drop=True)
    
    return df

def create_rt_model(cases_data, gen_int, reporting_delay):
    """Create the PyMC model for Rt estimation"""
    
    n_days = len(cases_data)
    cases = cases_data['cases'].values
    day_of_week = cases_data['day_of_week'].values - 1  # Convert to 0-6
    
    # Parameters for splines
    n_knots_rt = max(10, n_days // 15)  # Knots for Rt spline
    n_knots_asc = max(8, n_days // 20)   # Knots for ascertainment spline
    
    knots_rt = np.linspace(0, n_days-1, n_knots_rt)
    knots_asc = np.linspace(0, n_days-1, n_knots_asc)
    
    with pm.Model() as model:
        
        # ==================== Rt MODEL ====================
        # Random walk for Rt on log scale
        log_rt_init = pm.Normal('log_rt_init', mu=0, sigma=0.5)
        log_rt_steps = pm.Normal('log_rt_steps', mu=0, sigma=0.1, shape=n_days-1)
        log_rt = pt.concatenate([[log_rt_init], log_rt_init + pt.cumsum(log_rt_steps)])
        rt = pm.Deterministic('rt', pt.exp(log_rt))
        
        # ==================== DAY-OF-WEEK EFFECTS ====================
        # Day-of-week effects (multiplicative, sum to 7)
        dow_raw = pm.Normal('dow_raw', mu=0, sigma=0.3, shape=7)
        dow_effects = pm.Deterministic('dow_effects', 7 * pt.softmax(dow_raw))
        
        # ==================== TIME-VARYING ASCERTAINMENT ====================
        # Smooth ascertainment using random walk
        logit_asc_init = pm.Normal('logit_asc_init', mu=-2, sigma=1)  # Start low
        logit_asc_steps = pm.Normal('logit_asc_steps', mu=0, sigma=0.05, shape=n_days-1)
        logit_asc = pt.concatenate([[logit_asc_init], logit_asc_init + pt.cumsum(logit_asc_steps)])
        ascertainment = pm.Deterministic('ascertainment', pm.math.sigmoid(logit_asc))
        
        # ==================== INFECTION DYNAMICS ====================
        # Initial infections (seeding period)
        seed_days = min(14, n_days // 4)
        log_seed_infections = pm.Normal('log_seed_infections', mu=np.log(50), sigma=1, shape=seed_days)
        seed_infections = pt.exp(log_seed_infections)
        
        # Compute infections using renewal equation
        def compute_infections(rt_t, past_infections, gen_int):
            """Compute infections at time t using renewal equation"""
            # Convolve with generation interval
            infectiousness = 0
            for s in range(len(gen_int)):
                if s < past_infections.shape[0]:
                    infectiousness += past_infections[-(s+1)] * gen_int[s]
            return rt_t * infectiousness
        
        # Initialize infections list
        infections_list = [seed_infections[i] for i in range(seed_days)]
        
        # Compute remaining infections
        for t in range(seed_days, n_days):
            # Get past infections (up to generation interval length)
            past_infections = pt.stack(infections_list[max(0, t-len(gen_int)):t])
            
            # Compute new infections
            new_infection = compute_infections(rt[t], past_infections, gen_int)
            infections_list.append(new_infection)
        
        infections = pm.Deterministic('infections', pt.stack(infections_list))
        
        # ==================== OBSERVATION MODEL ====================
        # Apply reporting delay
        max_delay = len(reporting_delay)
        expected_reports = pt.zeros(n_days)
        
        for t in range(n_days):
            for d in range(max_delay):
                if t + d < n_days:
                    contribution = infections[t] * ascertainment[t] * reporting_delay[d]
                    expected_reports = pt.set_subtensor(
                        expected_reports[t + d], 
                        expected_reports[t + d] + contribution
                    )
        
        # Apply day-of-week effects
        dow_multiplier = dow_effects[day_of_week]
        expected_cases = expected_reports * dow_multiplier
        
        # ==================== LIKELIHOOD ====================
        # Overdispersion parameter
        phi = pm.Exponential('phi', lam=0.1)
        
        # Negative binomial likelihood
        likelihood = pm.NegativeBinomial(
            'likelihood',
            mu=expected_cases,
            alpha=phi,
            observed=cases
        )
        
        # Store expected cases for diagnostics
        pm.Deterministic('expected_cases', expected_cases)
    
    return model

def fit_model(model, samples=2000, tune=1000, chains=4):
    """Fit the PyMC model"""
    with model:
        # Use NUTS sampler
        trace = pm.sample(
            draws=samples,
            tune=tune,
            chains=chains,
            cores=min(4, chains),
            return_inferencedata=True,
            random_seed=42,
            target_accept=0.95
        )
    
    return trace

def extract_estimates(trace):
    """Extract point estimates and credible intervals"""
    summary = az.summary(trace, hdi_prob=0.95)
    
    # Extract Rt estimates
    rt_vars = [var for var in summary.index if var.startswith('rt[')]
    rt_summary = summary.loc[rt_vars]
    
    # Extract day-of-week effects
    dow_vars = [var for var in summary.index if var.startswith('dow_effects[')]
    dow_summary = summary.loc[dow_vars]
    
    # Extract ascertainment
    asc_vars = [var for var in summary.index if var.startswith('ascertainment[')]
    asc_summary = summary.loc[asc_vars]
    
    return {
        'rt': rt_summary,
        'dow_effects': dow_summary,
        'ascertainment': asc_summary,
        'full_summary': summary
    }

def create_plots(data, trace, estimates):
    """Create comprehensive plots of results"""
    
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))
    
    dates = data['date']
    n_days = len(dates)
    
    # Plot 1: Rt over time
    ax1 = axes[0, 0]
    rt_mean = estimates['rt']['mean'].values
    rt_lower = estimates['rt']['hdi_2.5%'].values
    rt_upper = estimates['rt']['hdi_97.5%'].values
    
    ax1.plot(dates, rt_mean, 'b-', linewidth=2, label='Rt estimate')
    ax1.fill_between(dates, rt_lower, rt_upper, alpha=0.3, color='blue', label='95% HDI')
    ax1.axhline(y=1, color='red', linestyle='--', alpha=0.7, label='Rt = 1')
    ax1.set_ylabel('Reproduction Number (Rt)')
    ax1.set_title('Time-varying Reproduction Number')
    ax1.legend()
    ax1.grid(True, alpha=0.3)
    plt.setp(ax1.xaxis.get_majorticklabels(), rotation=45)
    
    # Plot 2: Observed vs Expected Cases
    ax2 = axes[0, 1]
    posterior = trace.posterior
    expected_cases = posterior['expected_cases'].mean(dim=['chain', 'draw']).values
    
    ax2.scatter(data['cases'], expected_cases, alpha=0.6, s=30)
    max_val = max(data['cases'].max(), expected_cases.max())
    ax2.plot([0, max_val], [0, max_val], 'r--', alpha=0.7)
    ax2.set_xlabel('Observed Cases')
    ax2.set_ylabel('Expected Cases')
    ax2.set_title('Model Fit: Observed vs Expected')
    ax2.grid(True, alpha=0.3)
    
    # Plot 3: Day-of-week Effects
    ax3 = axes[1, 0]
    dow_names = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
    dow_mean = estimates['dow_effects']['mean'].values
    dow_lower = estimates['dow_effects']['hdi_2.5%'].values
    dow_upper = estimates['dow_effects']['hdi_97.5%'].values
    
    x_pos = np.arange(len(dow_names))
    ax3.bar(x_pos, dow_mean, yerr=[dow_mean - dow_lower, dow_upper - dow_mean], 
            capsize=5, alpha=0.7, color='green')
    ax3.set_xticks(x_pos)
    ax3.set_xticklabels(dow_names)
    ax3.set_ylabel('Reporting Multiplier')
    ax3.set_title('Day-of-Week Effects')
    ax3.grid(True, alpha=0.3, axis='y')
    
    # Plot 4: Time-varying Ascertainment
    ax4 = axes[1, 1]
    asc_mean = estimates['ascertainment']['mean'].values
    asc_lower = estimates['ascertainment']['hdi_2.5%'].values
    asc_upper = estimates['ascertainment']['hdi_97.5%'].values
    
    ax4.plot(dates, asc_mean, 'g-', linewidth=2, label='Ascertainment')
    ax4.fill_between(dates, asc_lower, asc_upper, alpha=0.3, color='green', label='95% HDI')
    ax4.set_ylabel('Ascertainment Probability')
    ax4.set_title('Time-varying Ascertainment')
    ax4.legend()
    ax4.grid(True, alpha=0.3)
    plt.setp(ax4.xaxis.get_majorticklabels(), rotation=45)
    
    plt.tight_layout()
    plt.show()
    
    # Additional plot: Cases over time with model fit
    fig, ax = plt.subplots(1, 1, figsize=(12, 6))
    
    ax.scatter(dates, data['cases'], alpha=0.6, s=30, label='Observed cases', color='black')
    ax.plot(dates, expected_cases, 'r-', linewidth=2, label='Expected cases', alpha=0.8)
    ax.set_ylabel('Daily Cases')
    ax.set_title('COVID-19 Cases: Observed vs Model Fit')
    ax.legend()
    ax.grid(True, alpha=0.3)
    plt.setp(ax.xaxis.get_majorticklabels(), rotation=45)
    plt.tight_layout()
    plt.show()

def main():
    """Main execution function"""
    print("COVID-19 Rt Estimation with Complex Observation Processes")
    print("=" * 60)
    
    # Load data
    print("\n1. Loading data...")
    data = load_data()
    print(f"   Loaded {len(data)} days of case data")
    print(f"   Date range: {data['date'].min()} to {data['date'].max()}")
    print(f"   Total cases: {data['cases'].sum():,}")
    
    # Create generation interval and reporting delay
    print("\n2. Setting up epidemiological parameters...")
    gen_int = create_generation_interval()
    reporting_delay = create_reporting_delay()
    print(f"   Generation interval mean: {np.sum(gen_int * np.arange(1, len(gen_int)+1)):.1f} days")
    print(f"   Reporting delay mean: {np.sum(reporting_delay * np.arange(len(reporting_delay))):.1f} days")
    
    # Create model
    print("\n3. Building PyMC model...")
    model = create_rt_model(data, gen_int, reporting_delay)
    print(f"   Model created with {len(data)} time points")
    
    # Check model
    with model:
        print(f"   Model has {len(model.free_RVs)} free random variables")
    
    # Fit model
    print("\n4. Fitting model (this may take several minutes)...")
    try:
        trace = fit_model(model, samples=1000, tune=500, chains=2)  # Reduced for faster execution
        print("   Model fitting completed successfully!")
    except Exception as e:
        print(f"   Error during fitting: {e}")
        return
    
    # Check convergence
    print("\n5. Checking convergence...")
    rhat = az.rhat(trace)
    max_rhat = float(rhat.max()) if hasattr(rhat.max(), 'values') else rhat.max().values.max()
    print(f"   Maximum R-hat: {max_rhat:.3f}")
    if max_rhat > 1.1:
        print("   Warning: Some parameters may not have converged (R-hat > 1.1)")
    else:
        print("   Convergence looks good (all R-hat <= 1.1)")
    
    # Extract estimates
    print("\n6. Extracting estimates...")
    estimates = extract_estimates(trace)
    
    # Print summary statistics
    print("\n7. Summary Results:")
    print("-" * 30)
    
    rt_mean = estimates['rt']['mean'].values
    print(f"   Rt range: {rt_mean.min():.2f} - {rt_mean.max():.2f}")
    print(f"   Current Rt: {rt_mean[-1]:.2f}")
    
    dow_effects = estimates['dow_effects']['mean'].values
    dow_names = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
    print(f"   Day-of-week effects:")
    for i, (day, effect) in enumerate(zip(dow_names, dow_effects)):
        print(f"     {day}: {effect:.2f}")
    
    asc_mean = estimates['ascertainment']['mean'].values
    print(f"   Ascertainment range: {asc_mean.min():.1%} - {asc_mean.max():.1%}")
    print(f"   Final ascertainment: {asc_mean[-1]:.1%}")
    
    # Create plots
    print("\n8. Creating plots...")
    create_plots(data, trace, estimates)
    
    print("\nAnalysis completed successfully!")
    
    return {
        'data': data,
        'model': model,
        'trace': trace,
        'estimates': estimates
    }

if __name__ == "__main__":
    results = main()

